### PR TITLE
new `Buffer`

### PR DESCRIPTION
**Pull request by cyx-6**

### CODE REVIEW

File: python/tvm/script/builder/tir/var.py [hunk 1]
```diff
@@ -15,24 +15,102 @@
...
+from tvm.tir import Var, IntImm
+from tvm._ffi import register_object as _register_object
+from tvm.runtime import Object, DataType
 from . import _ffi_api
 
 
-class BufferProxy:
-    def __call__(
+@_register_object("script.builder.tir.Buffer")
+class Buffer_(Object):
```
Comment (junrushao) [inline python/tvm/script/builder/tir/var.py:26]: just curious - is the class name going to conflict with the `Buffer` variable defined later in this file?

Comment (cyx-6) [reply python/tvm/script/builder/tir/var.py:26]: yes, it calls itself recursively, until Python crashes

File: src/script/builder/tir/var.h [hunk 2]
```diff
@@ -26,16 +26,36 @@ namespace script {
...
+  void VisitAttrs(tvm::AttrVisitor* v) { v->Visit("buffer", &buffer); }
+
+  static constexpr const char* _type_key = "script.builder.tir.Buffer";
+  TVM_DECLARE_BASE_OBJECT_INFO(BufferNode, runtime::Object);
+
+ public:
+  tvm::tir::BufferLoad BufferLoad(Array<PrimExpr> indices);
+  tvm::tir::BufferStore BufferStore(PrimExpr value, Array<PrimExpr> indices);
+  tvm::tir::BufferRegion BufferRegion(Array<Range> region);
+  tvm::tir::Prefetch Prefetch(Array<Range> bounds);
```
Comment (junrushao) [inline src/script/builder/tir/var.h]: in C++, do we want to enable overloaded `operator[]` that dispatches to `BufferLoad` and `BufferRegion`?

File: src/script/builder/tir/var.h [hunk 3]
```diff
@@ -26,16 +26,36 @@ namespace script {
...
+                 String buffer_name = "buffer", Optional<tvm::tir::Var> data = NullOpt,
+                 Optional<Array<PrimExpr>> strides = NullOpt,
+                 Optional<PrimExpr> elem_offset = NullOpt, String storage_scope = "", int align = 0,
+                 int offset_factor = 0, String buffer_type_str = "",
+                 Optional<Array<IntImm>> axis_separators = NullOpt);
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(Buffer, ObjectRef, BufferNode);
+
+ public:
+  inline void EnterWithScope();
+  inline void ExitWithScope();
```
Comment (junrushao) [inline src/script/builder/tir/var.h]: we dont need these two method, do we?

File: src/script/builder/tir/var.h (hunk 2 above)
Comment (junrushao) [reply src/script/builder/tir/var.h]: also, given these member methods are already implemented outside, probably we dont need those member methods?

File: src/script/builder/tir/var.cc [hunk 4]
```diff
@@ -84,7 +103,22 @@ TVM_STATIC_IR_FUNCTOR(Namer, vtable)
...
 
-TVM_REGISTER_GLOBAL("script.builder.tir.Buffer").set_body_typed(Buffer);
+TVM_REGISTER_NODE_TYPE(BufferNode);
+TVM_REGISTER_GLOBAL("script.builder.tir.Buffer")
+    .set_body_typed([](Array<PrimExpr> shape, DataType dtype = DataType::Float(32),
+                       String buffer_name = "buffer", Optional<tvm::tir::Var> data = NullOpt,
+                       Optional<Array<PrimExpr>> strides = NullOpt,
+                       Optional<PrimExpr> elem_offset = NullOpt, String storage_scope = "",
+                       int align = 0, int offset_factor = 0, String buffer_type_str = "",
+                       Optional<Array<IntImm>> axis_separators = NullOpt) {
```
Comment (junrushao) [inline src/script/builder/tir/var.cc]: No need to specify default values in the registry - they will be discarded

File: src/script/builder/tir/var.h (hunk 3 above)
Comment (cyx-6) [reply src/script/builder/tir/var.h]: Oh, I forget to delete these two methods

**Pull request merged**